ssl/quic/json_enc.c

```diff
@@ -0,0 +1,721 @@
+        return 1;
+
+    if (num_bytes <= OSSL_NELEM(json->stack_small)) {
+        stack = json->stack_small;
+    } else {
+        if (json->stack == json->stack_small)
+            json->stack = NULL;
+
+        stack = OPENSSL_realloc(json->stack, num_bytes);
+        if (stack == NULL)
+            return 0;
+    }
+
+    json->stack         = stack;
+    json->stack_bytes   = num_bytes;
+    return 1;
+}
+
+/* Push one bit onto the stack. Returns 0 on allocation failure. */
+static int json_push(OSSL_JSON_ENC *json, unsigned int v)
+{
+    if (v > 1)
+        return 0;
+
+    if (json->stack_end_byte >= json->stack_bytes) {
+        size_t new_size
+            = (json->stack_bytes == 0)
+            ? OSSL_NELEM(json->stack_small)
+            : (json->stack_bytes * 2);
+
+        if (!json_ensure_stack_size(json, new_size))
+            return 0;
+
+        json->stack_bytes = new_size;
+    }
+
+    if (v > 0)
+        json->stack[json->stack_end_byte] |= (v << json->stack_end_bit);
+    else
+        json->stack[json->stack_end_byte] &= ~(1U << json->stack_end_bit);
+
+    json->stack_end_bit = (json->stack_end_bit + 1) % CHAR_BIT;
+    if (json->stack_end_bit == 0)
+        ++json->stack_end_byte;
+
+    return 1;
+}
+
+/*
+ * Pop a bit from the stack. Returns 0 if stack is empty. Use json_peek() to get
+ * the value before calling this.
+ */
+static int json_pop(OSSL_JSON_ENC *json)
+{
+    if (json->stack_end_byte == 0 && json->stack_end_bit == 0)
+        return 0;
+
+    if (json->stack_end_bit == 0) {
+        --json->stack_end_byte;
+        json->stack_end_bit = CHAR_BIT - 1;
+    } else {
+        --json->stack_end_bit;
+    }
+
+    return 1;
+}
+
+/*
+ * Returns the bit on the top of the stack, or -1 if the stack is empty.
+ */
+static int json_peek(OSSL_JSON_ENC *json)
+{
+    size_t obyte, obit;
+
+    obyte = json->stack_end_byte;
+    obit  = json->stack_end_bit;
+    if (obit == 0) {
+       if (obyte == 0)
+           return -1;
+
+        --obyte;
+        obit = CHAR_BIT - 1;
+    } else {
+        --obit;
+    }
+
+    return (json->stack[obyte] & (1U << obit)) != 0;
+}
+
+/*
+ * OSSL_JSON_ENC: Initialisation
+ * =============================
+ */
+
+enum {
+    STATE_PRE_KEY,
+    STATE_PRE_ITEM,
+    STATE_PRE_COMMA
+};
+
+static ossl_inline int in_ijson(const OSSL_JSON_ENC *json)
+{
+    return (json->flags & OSSL_JSON_FLAG_IJSON) != 0;
+}
+
+static ossl_inline int in_seq(const OSSL_JSON_ENC *json)
+{
+    return (json->flags & OSSL_JSON_FLAG_SEQ) != 0;
+}
+
+static ossl_inline int in_pretty(const OSSL_JSON_ENC *json)
+{
+    return (json->flags & OSSL_JSON_FLAG_PRETTY) != 0;
+}
+
+int ossl_json_init(OSSL_JSON_ENC *json, BIO *bio, uint32_t flags)
+{
+    memset(json, 0, sizeof(*json));
+    json->flags     = flags;
+    json->error     = 1;
+    if (!wbuf_init(&json->wbuf, bio, 4096))
+        return 0;
+
+    json->state = STATE_PRE_COMMA;
+    return ossl_json_reset(json);
```

Comment:
Why is this reset necessary? Haven't we already reset everything via the memset above?